### ISSUE TITLE
wrong key resolving for trunk table hasOneThrough/hasManyThrough

### ISSUE THREAD

**Issue by yurii-github** [laravel/framework#28576]
- Laravel Version: 5.7
- Database Driver & Version: mysql

### Description:
wrong key resolving for trunk table

### Steps To Reproduce:
```
class SomeItem {

    public function trunk()
    {
        return $this->belongsTo(Trunk::class, 'trunk_id', 'id');
    }

    public function category()
    {
        return $this->hasOneThrough(Category::class, Trunk::class, 'id', 'id' , 'id', 'product_category_id')
            ->select('category.id as category_id');
    }
}
```
So, if we run 
`SomeItem::find(13)->category`

```
 App\Models\Category {#3519
     category_id: 5,
     id: 13, <------ MUST BE '5', '13' is ID of Trunk record
   }
```

Schema

**some_item**:
- id = 13
- trunk_id = 31 <----- issue with this column, instead of searching on '31' it will search on '13'

**tunk**:
- id  = 31   <---------- issue with this column, never searches on '31', instead searches on '13' and overwrites ID '5' from category row to '13'
- product_category_id = 5

**category**:
- id = 5
- name = 'some name'

--------------

**Comment by staudenmeir** [comment]
Your relationship is incorrect, you aren't specifying the `trunk_id` column anywhere:

```php
public function category()
{
    return $this->hasOneThrough(Category::class, Trunk::class, 'id', 'id', 'trunk_id', 'product_category_id')
        ->select('category.id as category_id');
}
```

**Comment by yurii-github** [comment]
@staudenmeir 
my bad, you are right about trunk_id, but it doesn't fix the issue of ID replacement.
with proper trunk_id  it properly searches on trunk now, but it still replaces ID of category with proper trunk ID now

```
=> App\Models\Category {#3519
     category_id: 5, <-- correct category ID
     id: 31, <--- correct trunk ID, but must be '5'
   }
```

**Comment by staudenmeir** [comment]
The column replacement only happens in Laravel 5.7, right?

**Comment by yurii-github** [comment]
yes

i have tested 5.7 only with hasManyThrough and  backported hasOneThrough from 5.8

**Comment by staudenmeir** [comment]
The column replacement issue was fixed in Laravel 5.8 (#25812), but the fix can't be backported to 5.7 (breaking change).

You can use this package of mine for Laravel < 5.8: https://github.com/staudenmeir/belongs-to-through

**Comment by yurii-github** [comment]
i see. thanks for your help!
